### PR TITLE
Support repair materials counted by charges

### LINKED ISSUE #17370: Wrong metal counting. Cannot repair or improve items that require scrap metal
Lately there was introduced charges counting system for scrap metal and few other metal items. So probably this bug caused by counting scrap metal by charges, and when improving action required more than one piece of scrap metal - it says that I need 2 pieces, but have only one (which contains say 4 "charges").

### PR DESCRIPTION
**Pull request by mugling**
Fixes #17370

### CODE REVIEW

**Comment by Coolthulhu** [comment]
This `if( count by charges ) { consume_charges } else { consume_amount }` structure is repeated a number of times in the code, we should eventually find a way to get a general solution to that.

**Comment by mugling** [comment]
An alternative is adjusting the scale of `volume` so as to avoid the issue of needing to have `count_by_charges` items

**Comment by codemime** [comment]
> An alternative is adjusting the scale of volume so as to avoid the issue of needing to have count_by_charges items

What if `volume` is zero?

**Comment by mugling** [comment]
> What if volume is zero?

For almost all cases it shouldn't be

**Comment by Coolthulhu** [comment]
> An alternative is adjusting the scale of volume so as to avoid the issue of needing to have count_by_charges items

And have a bottle of gasoline contain 250 "drops" of gasoline?

**Comment by mugling** [comment]
@Coolthulhu yes, a scale of `1vol == 1ml`

**Comment by Coolthulhu** [comment]
Well that would be an extreme number of items. Plus, it would mean updating all containers to understand multiple item content.

File: src/iuse_actor.cpp [hunk 1]
```diff
@@ -2236,9 +2236,12 @@ bool repair_item_actor::handle_components( player &pl, const item &fix,
     // Go through all discovered repair items and see if we have any of them available
     for( const auto &entry : valid_entries ) {
         const auto component_id = entry.obj().repaired_with();
-        if( crafting_inv.has_amount( component_id, items_needed ) ) {
-            // We've found enough of a material, add it to list
-            comps.push_back( item_comp( component_id, items_needed ) );
+        if( item::find_type( component_id )->count_by_charges() ) {
```
Comment (BevapDin) [inline src/iuse_actor.cpp]: Please use the static `item::count_by_charges` function, it takes the item type id as parameter.

**Comment by BevapDin** [comment]
`has_amount` and all the other amount function should handle items counted by charges just as well. The count-by-charges property should only mean one increments the counter by `the_item.charges` instead of `1`.

This is done correctly in `has_quality_internal` (visitable.cpp).

> And have a bottle of gasoline contain 250 "drops" of gasoline?

No thanks. This will only end up with a puddle of different gasoline items (same are damaged, some are cold, some have other flags). One body of liquid is one item.

Count-by-charges should ultimately be factored out in a `item_stack` class: it would contain the item representation (instance of `item`) and a count (`size_t`) and would represent N identical items (just like count-by-charges does now). But this would easily support 3 identical (to the tiniest detail) gun items and would get rid of the manual stacking code in various places (pickup, inventory, AIM).

File: src/iuse_actor.cpp (hunk 1 above)
Comment (mugling) [inline src/iuse_actor.cpp]: Done

File: src/iuse_actor.cpp [hunk 2]
```diff
@@ -2249,7 +2252,10 @@ bool repair_item_actor::handle_components( player &pl, const item &fix,
                 pl.add_msg_if_player( m_info,
                     _("You don't have enough %s to do that. Have: %d, need: %d"),
                     item::nname( mat_comp, 2 ).c_str(),
-                    crafting_inv.amount_of( mat_comp, false ), items_needed );
+                    item::find_type( mat_comp )->count_by_charges() ?
```
Comment (Coolthulhu) [inline src/iuse_actor.cpp]: Also here

**Pull request merged**